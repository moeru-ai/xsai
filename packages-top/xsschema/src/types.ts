import type { StandardSchemaV1 } from '@standard-schema/spec'
import type { JSONSchema7 } from 'json-schema'

export type { StandardSchemaV1 as Schema } from '@standard-schema/spec'
export type { JSONSchema7 as JSONSchema } from 'json-schema'

export type Infer<T extends StandardSchemaV1> = StandardSchemaV1.InferOutput<T>
export type InferIn<T extends StandardSchemaV1> = StandardSchemaV1.InferInput<T>

export type StandardSchemaToJsonSchemaFunction = (schema: StandardSchemaV1) => JSONSchema7

export type StandardSchemaVendor = 'arktype' | 'effect' | 'valibot' | 'zod'
