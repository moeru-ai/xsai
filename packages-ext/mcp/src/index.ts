export * from './utils/get-tools'
