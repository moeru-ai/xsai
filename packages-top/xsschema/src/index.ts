export { toJsonSchema, toJsonSchema as toJSONSchema, toJsonSchemaSync, toJsonSchemaSync as toJSONSchemaSync } from './to-json-schema'
export type { Infer, InferIn, JSONSchema as JsonSchema, JSONSchema, Schema, StandardSchemaVendor } from './types'
export { validate } from './validate'
export { registerStandardSchemaVendor, registerStandardSchemaVendors } from './vendors'
