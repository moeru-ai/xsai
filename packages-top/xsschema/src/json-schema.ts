import type { JsonSchema } from './types'

export const jsonSchema = (schema: JsonSchema) => schema

export const strictJsonSchema = (schema: JsonSchema): JsonSchema => ({
  ...schema,
  additionalProperties: false,
  properties: schema.properties
    ? Object.fromEntries(
        Object.entries(schema.properties)
          .map(([k, v]) => [
            k,
            typeof v === 'object' && 'type' in v && v.type === 'object'
              ? strictJsonSchema(v)
              : v,
          ]),
      )
    : schema.properties,
})
