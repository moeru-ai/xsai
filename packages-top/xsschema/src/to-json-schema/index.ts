import type { StandardSchemaV1 } from '@standard-schema/spec'
import type { JSONSchema7 } from 'json-schema'

export const toJsonSchema = async (schema: StandardSchemaV1): Promise<JSONSchema7> => {
  switch (schema['~standard'].vendor) {
    case 'arktype':
      return (await ((await import('./arktype')).toJsonSchema()))(schema)
    case 'effect':
      return (await ((await import('./effect')).toJsonSchema()))(schema)
    case 'valibot':
      return (await ((await import('./valibot')).toJsonSchema()))(schema)
    case 'zod':
      return (await ((await import('./zod')).toJsonSchema()))(schema)
    default:
      throw new Error(`xsschema: Unsupported schema vendor "${schema['~standard'].vendor}"`)
  }
}
