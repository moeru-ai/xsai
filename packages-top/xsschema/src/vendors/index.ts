import type { StandardSchemaToJsonSchemaFunction, StandardSchemaVendor } from '../types'

const vendorRegistryToJsonSchemaFunctions: Partial<Record<StandardSchemaVendor, StandardSchemaToJsonSchemaFunction | undefined>> = { }

/**
 * Registers a third-party vendor for converting Standard Schemas to JSON schemas.
 *
 * It is idempotent, so feel free to call it multiple times, but it really only
 * has to be called once per vendor.
 */
export const registerStandardSchemaVendor = async (vendor: StandardSchemaVendor): Promise<void> => {
  if (vendorRegistryToJsonSchemaFunctions[vendor])
    return

  switch (vendor) {
    case 'arktype':
      vendorRegistryToJsonSchemaFunctions.arktype = (await import('./arktype')).toJsonSchema
      break
    case 'effect':
      vendorRegistryToJsonSchemaFunctions.effect = (await import('./effect')).toJsonSchema
      break
    case 'valibot':
      vendorRegistryToJsonSchemaFunctions.valibot = (await import('./valibot')).toJsonSchema
      break
    case 'zod':
      vendorRegistryToJsonSchemaFunctions.zod = (await import('./zod')).toJsonSchema
      break
    default:
      throw new Error(`xsschema: Unsupported schema vendor "${vendor as string}"`)
  }
}

/**
 * Registers all installed third-party standard schema vendors.
 *
 * It is idempotent, so feel free to call it multiple times, but it really only
 * has to be called once.
 */
export const registerStandardSchemaVendors = async (): Promise<void> => {
  await Promise.all([
    registerStandardSchemaVendor('arktype').catch(),
    registerStandardSchemaVendor('effect').catch(),
    registerStandardSchemaVendor('valibot').catch(),
    registerStandardSchemaVendor('zod').catch(),
  ])
}

export const getVendorToJsonSchemaFunction = (vendor: StandardSchemaVendor): StandardSchemaToJsonSchemaFunction | undefined => {
  return vendorRegistryToJsonSchemaFunctions[vendor]
}
