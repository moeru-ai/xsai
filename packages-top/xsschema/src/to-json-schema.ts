import type { StandardSchemaV1 } from '@standard-schema/spec'
import type { JSONSchema7 } from 'json-schema'

import type { StandardSchemaVendor } from './types'

import { getVendorToJsonSchemaFunction, registerStandardSchemaVendor } from './vendors'

/**
 * Converts a Standard Schema to a JSON schema synchronously.
 *
 * @note Make sure to call `registerStandardSchemaVendor` with the appropriate vendor before calling this function, or you can call `toJsonSchema` instead which is async.
 */
export const toJsonSchemaSync = (schema: StandardSchemaV1): JSONSchema7 => {
  const { vendor } = schema['~standard']
  const toJsonSchema = getVendorToJsonSchemaFunction(vendor as StandardSchemaVendor)
  if (!toJsonSchema) {
    throw new Error(`xsschema: Unregistered or unsupported schema vendor "${vendor}". Make sure to register the vendor using "await registerStandardSchemaVendor('${vendor}')" before calling toJsonSchema.`)
  }

  return toJsonSchema(schema)
}

/**
 * Converts a Standard Schema to a JSON schema.
 *
 * @note This method is `async` because it has to `await import` the schema vendor's dependencies.
 */
export const toJsonSchema = async (schema: StandardSchemaV1): Promise<JSONSchema7> => {
  const { vendor } = schema['~standard']
  await registerStandardSchemaVendor(vendor as StandardSchemaVendor)
  return toJsonSchemaSync(schema)
}
