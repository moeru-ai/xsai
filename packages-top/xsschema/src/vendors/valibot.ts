import type { StandardSchemaV1 } from '@standard-schema/spec'
import type { JSONSchema7 } from 'json-schema'
import type { BaseIssue, BaseSchema } from 'valibot'

import { toJsonSchema as toJsonSchemaImpl } from '@valibot/to-json-schema'

export const toJsonSchema = (schema: StandardSchemaV1): JSONSchema7 => {
  return toJsonSchemaImpl(schema as BaseSchema<unknown, unknown, BaseIssue<unknown>>)
}
