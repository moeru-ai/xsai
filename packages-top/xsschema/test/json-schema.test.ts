import * as v from 'valibot'
import { describe, expect, it } from 'vitest'

import { strictJsonSchema, toJsonSchema } from '../src'

describe('strictJsonSchema', () => {
  it('basic', async () => {
    const schema = v.pipe(
      v.object({
        myString: v.string(),
        myUnion: v.union([v.number(), v.boolean()]),
      }),
      v.description('My neat object schema'),
    )

    expect(strictJsonSchema(await toJsonSchema(schema))).toMatchSnapshot()
  })

  it('nested', async () => {
    const schema = v.object({
      foo: v.object({
        bar: v.string(),
        baz: v.object({
          qux: v.string(),
        }),
      }),
    })

    expect(strictJsonSchema(await toJsonSchema(schema))).toMatchSnapshot()
  })
})
