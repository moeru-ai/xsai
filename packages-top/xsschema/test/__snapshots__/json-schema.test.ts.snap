// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`strictJsonSchema > basic 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "description": "My neat object schema",
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "anyOf": [
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;

exports[`strictJsonSchema > nested 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "foo": {
      "additionalProperties": false,
      "properties": {
        "bar": {
          "type": "string",
        },
        "baz": {
          "additionalProperties": false,
          "properties": {
            "qux": {
              "type": "string",
            },
          },
          "required": [
            "qux",
          ],
          "type": "object",
        },
      },
      "required": [
        "bar",
        "baz",
      ],
      "type": "object",
    },
  },
  "required": [
    "foo",
  ],
  "type": "object",
}
`;
